### DISCUSSION TITLE
Allow for completely overriding (and/or disabling) the core `dark:` variant.

### DISCUSSION THREAD

**Discussion by MichaelAllenWarner** [tailwindlabs/tailwindcss#11534] [Ideas]
I think it should be possible to write a completely custom `dark:` variant, but right now `addVariant('dark', /* ... */)` does nothing (presumably because the core `dark:` definition is applied after custom variants are registered).

My present motivation is making a "progressively enhanced" dark mode, where the base mechanism relies on `@media (prefers-color-scheme: dark)`, but where an attribute/class-based mechanism (with higher precedence) is added on top of that if JavaScript is enabled. The trouble is that the available configuration options for `dark:` aren't up to the task, and I can't override the core `dark:` variant with custom selectors. I _could_ define a separate variant like `dark-mode:` that does what I want, but since I can't disable `dark:`, developers are bound to accidentally use the latter instead of the former.

If the TW team were amenable to the idea, I might be able to find time to contribute here.

**Comment by MichaelAllenWarner** [comment]
If I'm not mistaken, simply moving `variantPlugins['darkVariants']` from `afterVariants` to `beforeVariants` [here](https://github.com/tailwindlabs/tailwindcss/blob/37575ea0bd573a96d10f3ba4d063020abc7c5825/src/lib/setupContextUtils.js#L744-L765) would allow for a fully custom `addVariant('dark', /* ... */)`. Is that right? And if so, is there any compelling reason not to?

**Comment by MichaelAllenWarner** [comment]
Another thought is something like what I've proposed [here](https://github.com/tailwindlabs/tailwindcss/discussions/11613) for the `aria-` variant definitions: add an _advanced_ option for the `dark:` variant setting, where the user can supply an entirely custom selector or array of selectors with the `&` syntax (as if they were supplying the second parameter of an `addVariant()` plugin, which is indeed what's happening [in the code](https://github.com/tailwindlabs/tailwindcss/blob/master/src/corePlugins.js#L227-L231)).

**Comment by olets** [comment]
Already possible: set `darkMode: null` and then add your custom variant

https://play.tailwindcss.com/NrRJTh1bQn?file=config

**Reply by MichaelAllenWarner** [reply]
That's terrific—many thanks!

**Reply by dario-baumberger** [reply]
Or `darkMode: "false"` but it's not mutch better...

**Reply by olets** [reply]
> Or `darkMode: "false"`

@dario-baumberger in my experience that's a type error and doesn't enable a custom dark mode variant https://play.tailwindcss.com/yZ7bzW94Sq?file=config

**Reply by dario-baumberger** [reply]
You're right. I didn't notice in my local stack because it's JS not TS. So null is the only correct solution

**Reply by olets** [reply]
[Tailwind 3.4.1](https://github.com/tailwindlabs/tailwindcss/releases/tag/v3.4.1) added true support for this via the "variant" dark mode strategy 🎉 

```js
darkMode: ["variant", ['@media (prefers-color-scheme: dark) { & }']]
```

https://play.tailwindcss.com/wBTUzgxVYF

**Reply by VinceBT-BG** [reply]
Ty @olets, works greatly with `darkMode: ['variant', [':is(.dark &):not(.light &)']]` !

**Reply by zimoo354** [reply]
@VinceBT-BG how did you achieved it? I cannot make it work 😕

**Reply by MichaelAllenWarner** [reply]
@zimoo354 It's documented here (but you need to be using Tailwind 3.4.1+): https://tailwindcss.com/docs/dark-mode#overriding-the-dark-variant
